### ISSUE TITLE
Using @* in xpath is throwing a error message in 1.22.1 release

### ISSUE THREAD

**Issue by yadavsvmx** [appium/appium#16142]
## The problem

Using @* in xpath is throwing a error message in 1.22.1 release
e.g:
 //*[@*='Hello']
 //*[contains(@*,'Hello')]
 (//*[contains(@*,'Hello')])[last()]

<img width="1539" alt="Screenshot 2021-11-22 at 12 18 17 PM" src="https://user-images.githubusercontent.com/30145363/142816726-d25b3de4-e964-4b24-b5f7-d100c5726e4e.png">

## Environment

* Appium version (or git revision) that exhibits the issue: 1.22.1
* Last Appium version that did not exhibit the issue (if applicable): 1.22
* Desktop OS/version used to run Appium: NA
* Node.js version (unless using Appium.app|exe): NA
* Npm or Yarn package manager: NA
* Mobile platform/version under test: Android 11 / ISO 15
* Real device or emulator/simulator: Real device
* Appium CLI or Appium.app|exe:

## Details

If necessary, describe the problem you have been experiencing in more detail.

## Link to Appium logs

Create a [GIST](https://gist.github.com) which is a paste of your _full_ Appium logs, and link them here.
Do _NOT_ paste your full Appium logs here, as it will make this issue very long and hard to read!
If you are reporting a bug, _always_ include Appium logs!

## Code To Reproduce Issue [ Good To Have ]

Please remember that with sample code it's easier to reproduce the bug and it's much faster to fix it.

Please git clone https://github.com/appium/appium and from the `sample-code` directory, use one of your favourite languages and sample apps to reproduce the issue.

In case a similar scenario is missing in sample-code, please submit a PR with one of the sample apps provided.

**Comment by mykola-mokhnach** [comment]
Please attach the server log

**Comment by yadavsvmx** [comment]
Attaching logs:
https://gist.github.com/yadavsvmx/48ceeaf4034440735ee5eabffb5b0d66

**Comment by mykola-mokhnach** [comment]
Could you also provide the actual XML page source?

**Comment by mykola-mokhnach** [comment]
I assume the issue might be the switch to XPath 2.0, where the behavior of `contains` function has been slightly changed. Check https://developers.perfectomobile.com/display/TT/XPath+2.0+Tips for more details.

**Comment by yadavsvmx** [comment]
But there re many xpaths already written in this style for existing projects, so it should support backward compatibility.

Else this will cascade into huge maintenance efforts

**Comment by mykola-mokhnach** [comment]
I don't think there is an easy way to ensure backward compatibility. Also, complex xpath locators are not being used very frequently as for my experience. We'd need to gather more feedback there to make particular conclusions.

**Comment by jlipps** [comment]
This is a good example of why it's beneficial to switch to Appium 2.0, so that you can choose to upgrade the android driver or not without it affecting upgrade paths for Appium or other drivers.

**Comment by KazuCocoa** [comment]
Oh, i see... We should address this in https://github.com/appium/appium/blob/master/CHANGELOG.md . Will do it later.

**Comment by mykola-mokhnach** [comment]
Closed as third party issue

**Comment by vivek-git-checkout** [comment]
> I don't think there is an easy way to ensure backward compatibility. Also, complex xpath locators are not being used very frequently as for my experience. We'd need to gather more feedback there to make particular conclusions.

I am also facing same error "java.lang.ClassCastException: java.util.ArrayList$ListItr cannot be cast to org.eclipse.wst.xml.xpath2.processor.internal.types.NodeTypeBuild" for xpath 
(//android.widget.TextView[@text='some, text']/following::android.widget.ImageButton)[1] with appium 1.22.2 which used to work till 1.22.0. Can someone please give right syntax for this xpath which would work with xpath2.0 ( @igorgiumellizup same issue as https://github.com/appium/appium/issues/16145)? Appium Inspector doesn't give any error for it, but we get it when we run test with automation.
Also, we have many such xpaths in our automation code. Will have to update all of them!?

**Comment by vivek-git-checkout** [comment]
> 

[page_source_for_appium_1.22.2_issue.txt](https://github.com/appium/appium/files/7800509/page_source_for_appium_1.22.2_issue.txt)

Created below alternate xpath for same element which worked with 1.22.2:
```
//android.widget.TextView[@text='some, text']/ancestor::android.widget.RelativeLayout//android.widget.ImageButton
```
Do we need to do this for all incompatible xpaths or can we have other way to make them compatible (with some flag or something) with 1.22.1 onwards Appium releases?

**Comment by mykola-mokhnach** [comment]
@vivek-git-checkout Could you please provide the full log with the error stacktrace? This could be a bug though

**Comment by mykola-mokhnach** [comment]
I did several experiments locally and it looks like the issue lies in the Eclipse's xpath2 processor itself and is specific only to queries containing `following` or `preceding` axes. I've reported an [issue](https://bugs.eclipse.org/bugs/show_bug.cgi?id=578021) to them. Lets see what the response would be. Maybe we'd need to have a temporary setting, which would allow to switch back to use xpath1 if we get no response from the Eclipse team soon.

**Comment by mykola-mokhnach** [comment]
Added a new driver setting called [enforceXPath1](https://github.com/appium/appium-uiautomator2-driver#settings-api), which allows to enforce XPath1 usage for element locators. This has been published since the driver version 1.75.0

**Comment by vivek-git-checkout** [comment]
> 

But now Xpath with lower-case is not working with enforceXPath1=true on Appium 1.23.0-beta.0 (appium-uiautomator2-driver version 1.75.0). Failing with "io.appium.uiautomator2.common.exceptions.InvalidSelectorException: javax.xml.xpath.XPathExpressionException: javax.xml.transform.TransformerException: Could not find function: lower-case" :

```
[HTTP] --> POST /wd/hub/session/37c262e3-599b-4da6-9df9-4f29312ce801/element
[HTTP] {"using":"xpath","value":"//android.widget.TextView[contains(lower-case(@text),'safety is our top priority')]"}
[debug] [W3C (37c262e3)] Calling AppiumDriver.findElement() with args: ["xpath","//android.widget.TextView[contains(lower-case(@text),'safety is our top priority')]","37c262e3-599b-4da6-9df9-4f29312ce801"]
[debug] [BaseDriver] Valid locator strategies for this request: xpath, id, class name, accessibility id, css selector, -android uiautomator
[debug] [BaseDriver] Waiting up to 0 ms for condition
[debug] [WD Proxy] Matched '/element' to command name 'findElement'
[debug] [WD Proxy] Proxying [POST /element] to [POST http://127.0.0.1:8200/wd/hub/session/4c065031-f346-4add-ae4d-61fa1f94de18/element] with body: {"strategy":"xpath","selector":"//android.widget.TextView[contains(lower-case(@text),'safety is our top priority')]","context":"","multiple":false}
[WD Proxy] Got response with status 400: {"sessionId":"4c065031-f346-4add-ae4d-61fa1f94de18","value":{"error":"invalid selector","message":"javax.xml.xpath.XPathExpressionException: javax.xml.transform.TransformerException: Could not find function: lower-case","stacktrace":"io.appium.uiautomator2.common.exceptions.InvalidSelectorException: javax.xml.xpath.XPathExpressionException: javax.xml.transform.TransformerException: Could not find function: lower-case\n\tat io.appium.uiautomator2.core.AccessibilityNodeInfoDumper.findNodesUsingXpath1(AccessibilityNodeInfoDumper.java:220)\n\tat io.appium.uiautomator2.core.AccessibilityNodeInfoDumper.findNodes(AccessibilityNodeInfoDumper.java:334)\n\tat io.appium.uiautomator2.utils.ElementLocationHelpers.getXPathNodeMatch(ElementLocationHelpers.java:119)\n\tat io.appium.uiautomator2.handler.FindElement.findElement(FindElement.java:88)\n\tat io.appium.uiautomator2.handler.FindElement.safeHandle(FindElement.java:67)\n\tat io.appium.uiautomator2.handler.request.SafeRequestHandler.handle(SafeRequestHandler.java:59...
[debug] [W3C] Matched W3C error code 'invalid selector' to InvalidSelectorError
[debug] [W3C (37c262e3)] Encountered internal error running command: io.appium.uiautomator2.common.exceptions.InvalidSelectorException: javax.xml.xpath.XPathExpressionException: javax.xml.transform.TransformerException: Could not find function: lower-case
[debug] [W3C (37c262e3)] 	at io.appium.uiautomator2.core.AccessibilityNodeInfoDumper.findNodesUsingXpath1(AccessibilityNodeInfoDumper.java:220)
[debug] [W3C (37c262e3)] 	at io.appium.uiautomator2.core.AccessibilityNodeInfoDumper.findNodes(AccessibilityNodeInfoDumper.java:334)
[debug] [W3C (37c262e3)] 	at io.appium.uiautomator2.utils.ElementLocationHelpers.getXPathNodeMatch(ElementLocationHelpers.java:119)
[debug] [W3C (37c262e3)] 	at io.appium.uiautomator2.handler.FindElement.findElement(FindElement.java:88)
[debug] [W3C (37c262e3)] 	at io.appium.uiautomator2.handler.FindElement.safeHandle(FindElement.java:67)
[debug] [W3C (37c262e3)] 	at io.appium.uiautomator2.handler.request.SafeRequestHandler.handle(SafeRequestHandler.java:59)
[debug] [W3C (37c262e3)] 	at io.appium.uiautomator2.server.AppiumServlet.handleRequest(AppiumServlet.java:267)
[debug] [W3C (37c262e3)] 	at io.appium.uiautomator2.server.AppiumServlet.handleHttpRequest(AppiumServlet.java:261)
[debug] [W3C (37c262e3)] 	at io.appium.uiautomator2.http.ServerHandler.channelRead(ServerHandler.java:68)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:366)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:352)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:345)
[debug] [W3C (37c262e3)] 	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:366)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:352)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:345)
[debug] [W3C (37c262e3)] 	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:435)
[debug] [W3C (37c262e3)] 	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293)
[debug] [W3C (37c262e3)] 	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:267)
[debug] [W3C (37c262e3)] 	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:250)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:366)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:352)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:345)
[debug] [W3C (37c262e3)] 	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:266)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:366)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:352)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:345)
[debug] [W3C (37c262e3)] 	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1294)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:366)
[debug] [W3C (37c262e3)] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:352)
[debug] [W3C (37c262e3)] 	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:911)
[debug] [W3C (37c262e3)] 	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:131)
[debug] [W3C (37c262e3)] 	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:611)
[debug] [W3C (37c262e3)] 	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:552)
[debug] [W3C (37c262e3)] 	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:466)
[debug] [W3C (37c262e3)] 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:438)
[debug] [W3C (37c262e3)] 	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:140)
[debug] [W3C (37c262e3)] 	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
[debug] [W3C (37c262e3)] 	at java.lang.Thread.run(Thread.java:764)
[debug] [W3C (37c262e3)] Caused by: javax.xml.xpath.XPathExpressionException: javax.xml.transform.TransformerException: Could not find function: lower-case
[debug] [W3C (37c262e3)] 	at org.apache.xpath.jaxp.XPathImpl.compile(XPathImpl.java:408)
[debug] [W3C (37c262e3)] 	at io.appium.uiautomator2.core.AccessibilityNodeInfoDumper.findNodesUsingXpath1(AccessibilityNodeInfoDumper.java:218)
[debug] [W3C (37c262e3)] 	... 38 more
[debug] [W3C (37c262e3)] Caused by: javax.xml.transform.TransformerException: Could not find function: lower-case
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.error(XPathParser.java:610)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.FunctionCall(XPathParser.java:1507)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.PrimaryExpr(XPathParser.java:1446)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.FilterExpr(XPathParser.java:1345)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.PathExpr(XPathParser.java:1278)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.UnionExpr(XPathParser.java:1236)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.UnaryExpr(XPathParser.java:1142)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.MultiplicativeExpr(XPathParser.java:1063)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.AdditiveExpr(XPathParser.java:1005)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.RelationalExpr(XPathParser.java:930)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.EqualityExpr(XPathParser.java:870)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.AndExpr(XPathParser.java:834)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.OrExpr(XPathParser.java:807)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.Expr(XPathParser.java:790)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.Argument(XPathParser.java:1469)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.FunctionCall(XPathParser.java:1537)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.PrimaryExpr(XPathParser.java:1446)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.FilterExpr(XPathParser.java:1345)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.PathExpr(XPathParser.java:1278)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.UnionExpr(XPathParser.java:1236)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.UnaryExpr(XPathParser.java:1142)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.MultiplicativeExpr(XPathParser.java:1063)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.AdditiveExpr(XPathParser.java:1005)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.RelationalExpr(XPathParser.java:930)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.EqualityExpr(XPathParser.java:870)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.AndExpr(XPathParser.java:834)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.OrExpr(XPathParser.java:807)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.Expr(XPathParser.java:790)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.PredicateExpr(XPathParser.java:1954)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.Predicate(XPathParser.java:1936)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.Step(XPathParser.java:1726)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.RelativeLocationPath(XPathParser.java:1626)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.LocationPath(XPathParser.java:1597)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.PathExpr(XPathParser.java:1317)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.UnionExpr(XPathParser.java:1236)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.UnaryExpr(XPathParser.java:1142)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.MultiplicativeExpr(XPathParser.java:1063)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.AdditiveExpr(XPathParser.java:1005)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.RelationalExpr(XPathParser.java:930)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.EqualityExpr(XPathParser.java:870)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.AndExpr(XPathParser.java:834)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.OrExpr(XPathParser.java:807)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.Expr(XPathParser.java:790)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.compiler.XPathParser.initXPath(XPathParser.java:129)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.XPath.<init>(XPath.java:178)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.XPath.<init>(XPath.java:266)
[debug] [W3C (37c262e3)] 	at org.apache.xpath.jaxp.XPathImpl.compile(XPathImpl.java:400)
[debug] [W3C (37c262e3)] 	... 39 more
[debug] [W3C (37c262e3)] 
[HTTP] <-- POST /wd/hub/session/37c262e3-599b-4da6-9df9-4f29312ce801/element 400 31 ms - 8310
```

**Comment by mykola-mokhnach** [comment]
Yes, previously xpath1 was the default engine and now it's xpath2. This is the reason I made it to a setting, you could switch it off only for buggy locators

**Comment by vivek-git-checkout** [comment]
> Yes, previously xpath1 was the default engine and now it's xpath2. This is the reason I made it to a setting, you could switch it off only for buggy locators

hmm, ok. but we have xpaths with following or preceding at 100+ places; same with lower-case 😟 lets see

**Comment by jlipps** [comment]
you could just set the setting as a capability once and it will be in effect for all your xpath queries.

**Comment by vivek-git-checkout** [comment]
> you could just set the setting as a capability once and it will be in effect for all your xpath queries.

Xpaths containing following or preceding axes are not working from Appium 1.22.1 onwards with xpath2.0. And xpath1.0 doesn't have "lower-case". We need both since we have 100+ xpaths with "following" and 100+ xpaths with "lower-case"; hence stuck here.

**Comment by cpignacio** [comment]
As per checking, using the latest uiautomator2 November 2024, this issue still persist.
Is there a workaround for me to use Xpath2 and Xpath1 at the same time?

**Comment by mykola-mokhnach** [comment]
XPath2 parser has been fixed in uia2 driver 7.5.2+
